### PR TITLE
Initial functionality

### PR DESCRIPTION
**Pull request by tijmenbaarda**
Initial version with support for HPB, VD16, VD17, VD18, Gallica, CERL Thesaurus and (limited) STCN. I tried to design it in such a way that parts can be reused for the VRE.

* I included STCN (Sparql/RDF) to show how the tool can work with very different types of APIs (SRU and Sparql endpoints), but this does not yet work as it should.
* Unfortunately Gallica and CERL Thesaurus have complex custom datatypes, so that I had to define custom ways to show their data to the users, but maybe there is a better solution.
* I tested on Windows and Linux. Installation on Windows with pip and git is straightforward.

### CODE REVIEW

**Review by jgonggrijp — CHANGES_REQUESTED** [review]
Solid work, well done! STCN does seem to work for me, although I needed a few attempts to find a query that yielded any results. "meilleur" gave me 209 results. Paging seems to work as intended as well.

Some things that stood out while trying, and which I think need fixing before merging:

- [x] It is not clear how to start the program. More detail in a comment below.
- [x] Inside the program, every prompt starts with `[1m# [0m` (at least on my mac), which looks weird. I'm guessing the intention is to show a nice prompt with markup and my mac is not having it. There is probably a more portable solution, or if it is difficult to fix, you could "downgrade" to a more boring plaintext prompt.

Things I think need fixing, but which can wait until after merging this branch:

- [ ] Handling of commands needs a more maintainable approach, but this should only be tackled after a first version of the tool has already been published so the users can start.
- [x] STCN results show all URIs in full, for example `http://schema.org/bookEdition`. I guess this is literally how it appears in the response body, but for the human reader it will be more convenient (and just as informative) when abbreviated, like `schema:bookEdition` in this case. I think this should *ideally* happen before publication of the first version, but it can also wait until the next version.
- [ ] `pytest` would be better than `unittest` (much nicer syntax and also more conventional within the lab). This should happen before adding more tests, regardless of whether this is before or after publication.

I did not spend much attention to the readers, but it looks nice and modular at first sight. Just one question: why are some readers inside the `readers` package, and some outside of it? Please let me know if there is any code in there that you would like me to spend more attention to.

Overall, I think it works quite well and I'm happy with the overall approach you have taken.

File: edpop_explorer/readers/cerl_thesaurus.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    data: dict = dataclass_field(default_factory=dict)
+    identifier: Optional[str] = None
+
+    def get_title(self) -> str:
+        if 'ctas:display' in self.data:
+            return self.data['ctas:display']
+        else:
+            return '(no display name defined)'
+
+    def show_dictlist(self, data: List[Dict[str, str]]) -> str:
```
[the remark this answers is not in the document]
Comment (tijmenbaarda) [reply edpop_explorer/readers/cerl_thesaurus.py:19]: I did not like this method either! I have replaced it by using a `pandas` `Dataframe`, which I already use in `GallicaReader`.

File: tests/test_srureader.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import json
+from pathlib import Path
+
+from edpop_explorer.srumarc21reader import SRUMarc21Reader
+
+
+TESTDATA = json.load(open(Path(__file__).parent / 'TESTDATA', 'r'))
+
+
+class SRUReaderTestCase(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (tijmenbaarda) [reply tests/test_srureader.py:12]: Will create an issue for this!

File: edpop_explorer/__main__.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            print('')
+            continue
+        if line.strip() != '':
+            try:
+                command, argument = line.split(maxsplit=1)
+            except ValueError:
+                command, argument = line, None
+        else:
+            continue
+        if command == 'exit':
```
[the remark this answers is not in the document]
Comment (tijmenbaarda) [reply edpop_explorer/__main__.py:134]: Sure, I will create an issue! Good to know about the `cmd` package, I had no idea that this existed!

**Comment by tijmenbaarda** [comment]
> I did not spend much attention to the readers, but it looks nice and modular at first sight. Just one question: why are some readers inside the `readers` package, and some outside of it? Please let me know if there is any code in there that you would like me to spend more attention to.
I have left the 'abstract' classes outside the `readers` package. I made a note about this in the readme file.

**Review by jgonggrijp — APPROVED** [review]
> Thanks for the review! I have addressed all your comments that you wanted to have fixed before approval/publication.

Well done!

> 
> Perhaps you could take a quick look at the way information is passed from `SparqlReader`'s subclasses to `SparqlRecord`. `SparqlRecord` needs to know `sparql_endpoint` and `prefixes`, but this information comes from `SparqlReader`'s subclasses. This information is now manually put into `SparqlRecord`, but maybe there is a better way.

As far as the endpoint is concerned, I don't think your current approach is too problematic. You could avoid some explicit management by also dedicating a SparqlRecord subclass to each corpus and setting the URL on that class just like you did for the reader. To keep it DRY, define a constant for the URL first, like this:

```py
STCN_ENDPOINT = 'http://data.bibliotheken.nl/sparql'


class STCNRecord(SparqlRecord):
    url = STCN_ENDPOINT
    # ...


class STCNReader(SparqlReader):
    url = STCN_ENDPOINT
    # ...
```

In case you need to share more attributes between classes, you can also define a third class that has all the common stuff and use it as a mixin or a member:

```py
# The common stuff
class STCNConfig(object):
    url = 'http://data.bibliotheken.nl/sparql'
    fancy = True

# Mixin variant
class STCNRecord(SparqlRecord, STCNConfig):
    # use self.url as before

# Member variant
class STCNRecord(SparqlRecord):
    # By not-so-widespread convention, "meta" or "Meta" is
    # sometimes used as a name for member classes.
    meta = STCNConfig
    # now use self.meta.url instead of self.url
```

As for the prefixes, I don't think those need to be specific to a corpus. You can just use the same set of namespaces for all corpora, which can be a constant. `SparqlRecord._replace_prefix` and `SparqlReader._give_sparql_prefixes` could just be standalone functions in the same module as the constant.

**Comment by tijmenbaarda** [comment]
> > Perhaps you could take a quick look at the way information is passed from `SparqlReader`'s subclasses to `SparqlRecord`. `SparqlRecord` needs to know `sparql_endpoint` and `prefixes`, but this information comes from `SparqlReader`'s subclasses. This information is now manually put into `SparqlRecord`, but maybe there is a better way.
> 
> As far as the endpoint is concerned, I don't think your current approach is too problematic. You could avoid some explicit management by also dedicating a SparqlRecord subclass to each corpus and setting the URL on that class just like you did for the reader. To keep it DRY, define a constant for the URL first, like this:

Thanks! Now that the prefixes are out of the classes as you suggested, it is only the `sparql_endpoint` and I will leave it like this for now.

**Pull request merged**